### PR TITLE
Improve Pylint parser

### PR DESCRIPTION
**Pull request by bpedersen2**
Adopt the parser to the standard 'parseable' output of pylint and use the symbolic message descriptor
instead of the numeric one.

This is a part of solving JENKINS-44915 and JENKINS-55340.  The changes to the warnings-ng plugin will follow later.

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java [hunk 1]
```diff
@@ -13,42 +16,59 @@
...
+    public static final String UNKNOWN_CAT = "pylint-unknown";
     /**
      * Creates a new instance of {@link PyLintParser}.
      */
     public PyLintParser() {
-        super(PYLINT_ERROR_PATTERN);
+        super(DEFAULT_PYLINT_ERROR_PATTERN);
+    }
+
+    public PyLintParser(@CheckForNull String regExp) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: What is the reason for this constructor?

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java [hunk 2]
```diff
@@ -13,42 +16,59 @@
...
     }
 
     @Override
     protected Optional<Issue> createIssue(final Matcher matcher, final IssueBuilder builder) {
-        String message = matcher.group(4);
-        String category = guessCategoryIfEmpty(matcher.group(3), message);
+        String message = matcher.group("message");
+        String category = matcher.group("symbol");
+        if (StringUtils.isEmpty(category)) {
+               category = UNKNOWN_CAT;
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: whitespace

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: wouldn't it make more sense to set the category to - in that case?

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java [hunk 3]
```diff
@@ -13,42 +16,59 @@
...
                 .buildOptional();
     }
 
-    private Severity mapPriorty(final String category) {
+    private Severity mapPriority(final String category) {
         //First letter of the Pylint classification is one of F/E/W/R/C. E/F/W are high priority.
         Severity priority;
 
         // See http://docs.pylint.org/output.html for definitions of the categories
+        if (category.isEmpty()) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: Better use `StringUtils` as `category` might be `null`

File: src/test/java/edu/hm/hafner/analysis/parser/PylintParserTest.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
  * Tests the class {@link PyLintParser}.
  */
 class PylintParserTest extends AbstractParserTest {
-    private static final String ISSUES_FILE = "pyLint.txt";
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/PylintParserTest.java]: Would make sense to support the old format as well, so add as additional test

**Comment by uhafner** [comment]
Are the files pymodule.py and init required?

File: pom.xml [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
   <groupId>edu.hm.hafner</groupId>
   <artifactId>analysis-model</artifactId>
-  <version>1.1.1-SNAPSHOT</version>
+  <version>1.2.0-SNAPSHOT</version>
```
Comment (uhafner) [inline pom.xml]: Should be not required?

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java [hunk 6]
```diff
@@ -13,42 +16,59 @@
...
  */
 public class PyLintParser extends FastRegexpLineParser {
     private static final long serialVersionUID = 4464053085862883240L;
 
-    private static final String PYLINT_ERROR_PATTERN = "(.*):(\\d+): \\[(\\D\\d*).*\\] (.*)";
+    // the default pattern matches "--output-format=parseable" output.
+    private static final String DEFAULT_PYLINT_ERROR_PATTERN = "(?<path>.*):(?<line>\\d+): \\[(?<category>\\D\\d*)\\((?<symbol>.*)\\), .*?\\] (?<message>.*)";
 
+
+    public static final String UNKNOWN_CAT = "pylint-unknown";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: package private

**Comment by uhafner** [comment]
Ping?

File: pom.xml (hunk 5 above)
Comment (bpedersen2) [reply pom.xml]: Rebased onto current master, so this is fixex

File: src/test/java/edu/hm/hafner/analysis/parser/PylintParserTest.java (hunk 4 above)
Comment (bpedersen2) [reply src/test/java/edu/hm/hafner/analysis/parser/PylintParserTest.java]: Done

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java (hunk 3 above)
Comment (bpedersen2) [reply src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: Done

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java (hunk 2 above)
Comment (bpedersen2) [reply src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: I prefer a more verbose one and pylint-specific here,

File: src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java (hunk 1 above)
Comment (bpedersen2) [reply src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java]: To be able to make the regex user-configurable. 

Now used in the legacy test, the changes for warnings-ng plugin will follow later.

**Comment by uhafner** [comment]
Are the files pymodule.py and init required?

**Comment by uhafner** [comment]
Hasn't the new format just a new additional (optional) category? So we can have one parser without a parameter to parse the old **and** the new format? So the LegacyTestCase would use the same parser, just some matcher groups are empty. Or is that not possible? How would a user choose the correct parser in the UI with your approach? Typically the user selects a parser and the parser is responsible to parse the format of different versions.

**Comment by bpedersen2** [comment]
> Hasn't the new format just a new additional (optional) category? So we can have one parser without a parameter to parse the old **and** the new format? So the LegacyTestCase would use the same parser, just some matcher groups are empty. Or is that not possible? How would a user choose the correct parser in the UI with your approach? Typically the user selects a parser and the parser is responsible to parse the format of different versions.

No, the order of the fields is quite different (and it is freely user-configurable on the pylint side). The idea is to supply a default that parses the 'default' (== 'parseable') output of pylint and allow the user to supply a suitable regex for other formats.

Actually, I have never seen a pylint version in the last 5 years that produces anything resembling the old testcase ( especially the 'C' only is somthing that  seems impossible to generate, as the msg_id is always [CRWE][0-9]{4}.

**Comment by uhafner** [comment]
Ok, thanks! I simplified the parser a little bit and remove the regexp in the API.

**Pull request merged**